sdk/identity/azure-identity/samples/chained_token_credential.cpp

```diff
@@ -0,0 +1,39 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include <iostream>
+
+#include <azure/identity/chained_token_credential.hpp>
+
+#include <azure/identity/environment_credential.hpp>
+#include <azure/identity/managed_identity_credential.hpp>
+
+#include <azure/service/client.hpp>
+
+int main()
+{
+  try
+  {
+    // Step 1: Initialize Chained Token Credential.
+    // A configuration demonstrated below would authenticate using EnvironmentCredential if it is
+    // available, and if it is not available, would fall back to use ManagedIdentityCredential.
+    auto chainedTokenCredential = std::make_shared<Azure::Identity::ChainedTokenCredential>(
+        Azure::Identity::ChainedTokenCredential::Sources{
+            std::make_shared<Azure::Identity::EnvironmentCredential>(),
+            std::make_shared<Azure::Identity::ManagedIdentityCredential>()});
```

Comment:
Should we tell customers that it might be a good idea to create a `TokenCredentialOptions` and use it for all the credentials within the chainedTokenCredential?  That way all credentials would share the same pipeline instead of allocating one pipeline per credential...

Reply:
Each of the credential has its own options class. So, when user initializes them, they may chose to pass different set of options, or the same one (it is allowed to pass just `TokenCredentialOptions`, and everything that is common will be reused.
Since options are taken at creation time, and `ChainedTokenCredential` takes a list of already created objects, it will not be possible to re-initialize already created object. But even if that was possible, I do not think it is a good idea. Each class has a constructor taking options - there is enough flexibility to pass them either same one, slightly customized, or separate. There is not place for `ChaniedTokenCredential` to have any options as of now. If in the future we want them for some reason (though not for the reasons described above - again, I don't think we should have them), then we can add these options to be the second parameter for the constructor with the default value, or an overload.

Reply:
Yeah, I see..
One of the cool things about some clients (like the storage client) is that they re-use the same pipeline and options across clients.  I think that would be nice for all the credentials inside the chained credential to use the same settings and pipeline (retry, transport, etc).  But there's also a benefit of letting each credential be unique.
For example, I can have one credential to use one transport adapter and another credential to use another transport adapter.
Ok